src/sundials/sundials_context.c

```diff
@@ -44,88 +62,105 @@ int SUNContext_Create(SUNComm comm, SUNContext* sunctx)
-#else
-  if (SUNLogger_CreateFromEnv(SUN_COMM_NULL, &logger))
-#endif
-  {
-    return (-1);
+  if (SUNLogger_CreateFromEnv(comm, &logger)) {
+    return SUN_ERR_LOGGER_CORRUPT;
   }
 #else
-#if SUNDIALS_MPI_ENABLED
-  if (SUNLogger_Create(comm, 0, &logger))
-#else
-  if (SUNLogger_Create(SUN_COMM_NULL, 0, &logger))
+  if (SUNLogger_CreateFromEnv(SUN_COMM_NULL, &logger)) {
+    return SUN_ERR_LOGGER_CORRUPT;
+  }
 #endif
-  {
-    return (-1);
+#else
+  if (SUNLogger_Create(NULL, 0, &logger)) {
+    return SUN_ERR_LOGGER_CORRUPT;
   }
-  SUNLogger_SetErrorFilename(logger, "");
-  SUNLogger_SetWarningFilename(logger, "");
-  SUNLogger_SetInfoFilename(logger, "");
-  SUNLogger_SetDebugFilename(logger, "");
+  SUNCheckCall(SUNLogger_SetErrorFilename(logger, ""));
+  SUNCheckCall(SUNLogger_SetWarningFilename(logger, ""));
+  SUNCheckCall(SUNLogger_SetInfoFilename(logger, ""));
+  SUNCheckCall(SUNLogger_SetDebugFilename(logger, ""));
 #endif
 
-  *sunctx = NULL;
-  *sunctx = (SUNContext)malloc(sizeof(struct _SUNContext));
-
-  if (*sunctx == NULL)
-  {
 #if defined(SUNDIALS_BUILD_WITH_PROFILING) && !defined(SUNDIALS_CALIPER_ENABLED)
-    SUNProfiler_Free(&profiler);
+  SUNCheckCall(SUNProfiler_Create(comm, "SUNContext Default", &profiler));
 #endif
-    SUNLogger_Destroy(&logger);
-    return (-1);
-  }
 
-  (*sunctx)->logger       = logger;
-  (*sunctx)->own_logger   = logger != NULL;
-  (*sunctx)->profiler     = profiler;
-  (*sunctx)->own_profiler = profiler != NULL;
+  sunctx->logger        = logger;
+  sunctx->own_logger    = logger != NULL;
+  sunctx->profiler      = profiler;
+  sunctx->own_profiler  = profiler != NULL;
+  sunctx->last_err      = 0;
+  sunctx->err_handler   = SUNErrHandler_Create(SUNLogErrHandlerFn, NULL);
+  sunctx->comm          = comm;
+
+  *sunctx_out = sunctx;
 
-  return (0);
+  return SUN_SUCCESS;
 }
 
-int SUNContext_GetProfiler(SUNContext sunctx, SUNProfiler* profiler)
+SUNErrCode SUNContext_GetLastError(SUNContext sunctx, SUNErrCode* last_err)
 {
-  if (sunctx == NULL)
+  *last_err = sunctx->last_err;
+  sunctx->last_err = SUN_SUCCESS;
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNContext_PeekLastError(SUNContext sunctx, SUNErrCode* last_err)
+{
+  *last_err = sunctx->last_err;
+  return SUN_SUCCESS;
+}
+
+SUNErrHandler SUNContext_PushErrHandler(SUNContext sunctx, SUNErrHandlerFn err_fn, void* err_user_data)
+{
+  SUNErrHandler new_err_handler = SUNErrHandler_Create(err_fn, err_user_data);
+  new_err_handler->previous = sunctx->err_handler;
+  sunctx->err_handler = new_err_handler;
+  return new_err_handler;
+}
+
+SUNErrCode SUNContext_PopErrHandler(SUNContext sunctx)
+{
+  if (sunctx->err_handler) {
+    SUNErrHandler eh = sunctx->err_handler;
+    if (sunctx->err_handler->previous) {
+      sunctx->err_handler = sunctx->err_handler->previous;
+    } else {
+      sunctx->err_handler = NULL;
+    }
+    free(eh);
+  }
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNContext_ClearHandlers(SUNContext sunctx)
```

Comment:
Ok, my concern was "clear" could be interpreted as clearing the errors rather than the stack but since we're keeping the `Get` and `Peek` functions and not adding a `Clear` function that confusion seems less likely.